nwss_wastewater/delphi_nwss/pull.py

```diff
@@ -0,0 +1,132 @@
+    # metric type conversion
+    signals_dict_metric = {key: float for key in METRIC_SIGNALS}
+    metric_dates_dict = {key: "datetime64[ns]" for key in METRIC_DATES}
+    type_dict_metric = {**metric_dates_dict, **signals_dict_metric, **SAMPLE_SITE_NAMES}
+    return type_dict, type_dict_metric
+
+
+def warn_string(df, type_dict):
+    """Format the warning string."""
+    return f"""
+Expected column(s) missed, The dataset schema may
+have changed. Please investigate and amend the code.
+
+Columns needed:
+{NEWLINE.join(type_dict.keys())}
+
+Columns available:
+{NEWLINE.join(df.columns)}
+"""
+
+
+def reformat(df, df_metric):
+    """Add the population column from df_metric to df, and rename some columns."""
+    # drop unused columns from df_metric
+    df_population = df_metric.loc[:, ["key_plot_id", "date_start", "population_served"]]
+    # get matching keys
+    df_population = df_population.rename(columns={"date_start": "timestamp"})
+    df_population = df_population.set_index(["key_plot_id", "timestamp"])
+    df = df.set_index(["key_plot_id", "timestamp"])
+
+    df = df.join(df_population)
+    df = df.reset_index()
+    return df
+
+
+def pull_nwss_data(token: str):
+    """Pull the latest NWSS Wastewater data, and conforms it into a dataset.
+
+    The output dataset has:
+
+    - Each row corresponds to (sewershed_key, day), denoted (geo_id, timestamp)
+    - Each row additionally has columns 'pcr_conc_smoothed', 'normalization',
+    'population_served', 'ptc_15d', 'detect_prop_15d',
+
+    Parameters
+    ----------
+    token: str
+        My App Token for pulling the NWSS data (could be the same as the nchs data)
+    test_file: Optional[str]
+        When not null, name of file from which to read test data
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe as described above.
+    """
+    # Constants
+    keep_columns = SIGNALS.copy()
+    # concentration key types
+    type_dict, type_dict_metric = construct_typedicts()
+
+    # Pull data from Socrata API
+    client = Socrata("data.cdc.gov", token)
+    results_concentration = client.get("g653-rqe2", limit=10**10)
+    results_metric = client.get("2ew6-ywp6", limit=10**10)
+    df_metric = pd.DataFrame.from_records(results_metric)
+    df = pd.DataFrame.from_records(results_concentration)
+    df = df.rename(columns={"date": "timestamp"})
+
+    try:
+        df = df.astype(type_dict)
+    except KeyError as exc:
+        raise ValueError(warn_string(df, type_dict)) from exc
+
+    try:
+        df_metric = df_metric.astype(type_dict_metric)
+    except KeyError as exc:
+        raise ValueError(warn_string(df_metric, type_dict_metric)) from exc
+
+    # pull 2 letter state labels out of the key_plot_id labels
+    df["state"] = df.key_plot_id.str.extract(r"_(\w\w)_")
+
+    # round out some of the numeric noise that comes from smoothing
+    for signal in SIGNALS:
+        df[signal] = sig_digit_round(df[signal], SIG_DIGITS)
```

Comment:
i know theres only one signal in `SIGNALS` right now, but do you intend to apply this to other signals going forward?  it might be needlessly deleterious or computationally expensive.

Reply:
Eventually I'll be doing this for `df_metric`, but that wouldn't need a for loop.